content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,386 @@
+
+This SIP is for now structured in three sections: 
+Motivation to visualize the desired outcome. 
+Desgin to enumarate all goals and constrains for this SIP.
+Encoding describe how the design can be implemented, focused on the encoding choosen in the PR.
+
+## Motivating
+
+Given one wants to use pattern matching on a case class:
+
+```scala
+case class User(name: String, age: Int, city: String)
+
+val user = User(name = "Anna", age = 10, city = "Berlin")
+
+val annasCity = user match
+  case User(name = "Anna", city = c) => c
+```
+
+The Deconstruction allows the same syntax as the construction and seems to be what people intuitively expect. See for example the very first post in [Scala Contributors Thread][contributors-thread] for this topic.
+
+Without names in patterns, users have to use underscore a lot. The example above would be written as:
+
+```scala
+val annasCity = user match
+  case User("Anna", _, c) => c
+```
+
+This common pattern makes it hard to see which parameter means what. The same issue led to named arguments. Adding underscores until the compiler is happy is also not a great experience. IDEs help here, by showing the names.
+
+Additionally, the code breaks every time a field of `User` gets added, rearranged, or removed.
+In the best case, the broken patterns fail to compile and needs to fixed manually.
+In the worst case, a pattern breaks silently, if two fields with the same type switch places.
+
+My personal motivation comes from using [`matchPattern` in ScalaTest](https://www.scalatest.org/user_guide/using_matchers#matchingAPattern)
+and got bitten by it every time my data model changed slightly. So named pattern matching is also allowed where ever normal pattern matching allowed:
+
+```scala
+user should matchPattern { case User(name = "Sarah") => } // ScalaTest
+
+val User(age = ageOfUser) = user
+
+for {
+  User(name = userName) <- db.fetchUser()
+} yield userName
+
+def neighboursNameOfUser(user: User, allUsers: List[User]) =
+  allUsers.collect {
+    case neighbour @ User(city = user.city) if neighbour != user => neighbour.name
+  }
+```
+
+## Design
+
+The goal is to allow named parameter in deconstruction as in construction.
+
+### Mixed usage
+
+Mixed patterns, with positional and named patterns are allowed to keep the similarity.
+But they have no motivational use case. Maybe they should be disallowed.
+
+```scala
+  case User("Anna", city = c) => // Mixed usage seems wired
+  case User(_, city = c) => // Leading underscore are especially useless
+```
+
+### Case classes with sequences
+
+For the case that the extractor is a mixure of [product and sequence match](https://dotty.epfl.ch/docs/reference/changed-features/pattern-matching.html#product-sequence-match) the name of the sequence can only be used at the last position.
+
+```
+case class Country(name: String, cities: String*)
+val Country(cities = "Berlin") = ???  // ok
```

Comment:
Would the following be valid as well?

~~~ scala
val Country(cities = "Berlin", "Hamburg") = ???
~~~

It is not clear that `"Hamburg"` would be part of the named field pattern, in my opinion.

I don’t really have a better suggestion (other than not supporting named field patterns in sequences), though.

Maybe the following?

~~~ scala
val Country(cities = Seq("Berlin", "Hamburg")*) = ???
~~~

But that looks heavy.

Reply:
I think it should be the first syntax, because the syntax is the same for construction.

```scala
val Country(cities = "Berlin", "Hamburg") =
  Country(name = "Germany", cities = "Berlin", "Hamburg")
```

I expanded the example. I played around the pattern below should be valid as well, for the same reason.

```scala
val Country(name = "Germany", "Berlin", "Hamburg") =
  Country(name = "Germany", "Berlin", "Hamburg")
```